x-pack/legacy/plugins/ml/public/application/jobs/components/custom_url_editor/utils.js

```diff
@@ -152,52 +156,42 @@ function buildDashboardUrlFromSettings(settings) {
           query = searchSourceData.query;
         }
 
-        // Add time settings to the global state URL parameter with $earliest$ and
-        // $latest$ tokens which get substituted for times around the time of the
-        // anomaly on which the URL will be run against.
-        const _g = rison.encode({
-          time: {
-            from: '$earliest$',
-            to: '$latest$',
-            mode: 'absolute',
-          },
-        });
-
-        const appState = {
-          filters,
-        };
-
-        // To put entities in filters section would involve creating parameters of the form
-        // filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:b30fd340-efb4-11e7-a600-0f58b1422b87,
-        // key:airline,negate:!f,params:(query:AAL,type:phrase),type:phrase,value:AAL),query:(match:(airline:(query:AAL,type:phrase)))))
-        // which includes the ID of the index holding the field used in the filter.
-
-        // So for simplicity, put entities in the query, replacing any query which is there already.
-        // e.g. query:(language:kuery,query:'region:us-east-1%20and%20instance:i-20d061fa')
         const queryFromEntityFieldNames = buildAppStateQueryParam(queryFieldNames);
         if (queryFromEntityFieldNames !== undefined) {
           query = queryFromEntityFieldNames;
         }
 
-        if (query !== undefined) {
-          appState.query = query;
-        }
-
-        const _a = rison.encode(appState);
-
-        const urlValue = `kibana#/dashboard/${dashboardId}?_g=${_g}&_a=${_a}`;
-
-        const urlToAdd = {
-          url_name: settings.label,
-          url_value: urlValue,
-          time_range: TIME_RANGE_TYPE.AUTO,
-        };
-
-        if (settings.timeRange.type === TIME_RANGE_TYPE.INTERVAL) {
-          urlToAdd.time_range = settings.timeRange.interval;
-        }
+        const generator = npStart.plugins.directAccessLinks.getAccessLinkGenerator(
+          DASHBOARD_APP_LINK_GENERATOR
+        );
+
+        return generator
+          .createUrl({
+            dashboardId,
+            timeRange: {
+              from: '$earliest$',
+              to: '$latest$',
+              mode: 'absolute',
+            },
+            filters,
```

Comment:
Not sure if/how this fits in with this PR, but one thing that would be useful would be for there to be an easy way to pass additional filter 'entities' to the target URL. Currently in ML when drilling down on an anomaly, say in 'airline:AAL', we pass this in the `query` part of the URL. Ideally we would pass this in the `filters` part of the URL, but we are using the `query` part as it is so difficult to construct the `filters` part of the URL correctly which needs to be in the form e.g.
```
filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:b30fd340-efb4-11e7-a600-0f58b1422b87,
 key:airline,negate:!f,params:(query:AAL,type:phrase),type:phrase,value:AAL),query:(match:(airline:(query:AAL,type:phrase)))))
```

It would be nice to have a way of saying 'drill down to this URL, adding airline=AAL, into the filters on the target page.

Reply:
I think what we just need is a helper service, something like `createFilter(fieldName: string, value: string): Filter`  Then you could use that and create an array of `Filter` objects to the URL generator.

cc @lizozom - do you know if there are any existing services for this?

Reply:
@stacey-gammon @lizozom yes some sort of helper service for generating the filters part of the URL would be great. As well as `createFilter`, I think we'd also want a function for adding a new filter to whatever filters are already set in the saved dashboard. 

For example, if the saved dashboard has one filter:

![image](https://user-images.githubusercontent.com/7405507/75546388-4fa6b480-5a20-11ea-944a-b2196299c721.png)

the drilldown might want to add an extra filter (here `instance: i-4fefbf6c`) to the filter which is in the saved object:

![image](https://user-images.githubusercontent.com/7405507/75546449-76fd8180-5a20-11ea-90da-b486dae60852.png)